src/solvers/refinement/string_constraint_generator_float.cpp

```diff
@@ -0,0 +1,454 @@
+      res.length(), ID_le, from_integer(j, res.length().type()));
+    mult_exprt ten_sum(sum, ten);
+
+    // sum = 10 * sum + after_end ? 0 : (res[j]-'0')
+    if_exprt to_add(
+      after_end,
+      from_integer(0, type),
+      typecast_exprt(minus_exprt(res[j], zero_char), type));
+    sum=plus_exprt(ten_sum, to_add);
+
+    and_exprt is_number(
+          binary_relation_exprt(res[j], ID_ge, zero_char),
+          binary_relation_exprt(res[j], ID_le, nine_char));
+    digit_constraints.push_back(is_number);
+
+    // There are no trailing zeros except for ".0" (i.e j=2)
+    if(j>1)
+    {
+      not_exprt no_trailing_zero(and_exprt(
+        equal_exprt(res.length(), from_integer(j+1, res.length().type())),
+      equal_exprt(res[j], zero_char)));
+      digit_constraints.push_back(no_trailing_zero);
+    }
+  }
+
+  exprt a2=conjunction(digit_constraints);
+  axioms.push_back(a2);
+
+  equal_exprt a3(int_expr, sum);
+  axioms.push_back(a3);
+
+  return res;
+}
+
+/// Add axioms to write the float in scientific notation.
+///
+/// A float is represented as $f = m * 2^e$ where $0 <= m < 2$ is the
+/// significand and $-126 <= e <= 127$ is the exponent.
+/// We want an alternate representation by finding $n$ and
+/// $d$ such that $f=n*10^d$. We can estimate $d$ the following way:
+/// $d ~= log_10(f/n) ~= log_10(m) + log_10(2) * e - log_10(n)$
+/// $d = floor(log_10(2) * e)$
+/// Then $n$ can be expressed by the equation:
+/// $log_10(n) = log_10(m) + log_10(2) * e - d$
+/// $n = f / 10^d = m * 2^e / 10^d = m * 2^e / 10^(floor(log_10(2) * e))$
+/// TODO: For now we only consider single precision.
+/// \param f: a float expression, which is positive
+/// \param max_size: a maximal size for the string
+/// \param ref_type: a type for refined strings
+/// \return a string expression representing the float in scientific notation
+string_exprt string_constraint_generatort::
+  add_axioms_from_float_scientific_notation(
+    const exprt &f, const refined_string_typet &ref_type)
+{
+  ieee_float_spect float_spec=ieee_float_spect::single_precision();
+  typet float_type=float_spec.to_type();
+  signedbv_typet int_type(32);
+
+  // This is used for rounding float to integers.
+  exprt round_to_zero_expr=from_integer(ieee_floatt::ROUND_TO_ZERO, int_type);
+
+  // `bin_exponent` is $e$ in the formulas
+  exprt bin_exponent=get_exponent(f, float_spec);
+
+  // $m$ from the formula is a value between 0.0 and 2.0 represented
+  // with values in the range 0x000000 0xFFFFFF so 1 corresponds to 0x800000.
+  // `bin_significand_int` represents $m * 0x800000$
+  exprt bin_significand_int=
+    get_significand(f, float_spec, unsignedbv_typet(32));
+  // `bin_significand` represents $m$ and is obtained
+  // by multiplying `binary_significand_as_int` by 1/0x800000=1.192092896e-7.
+  exprt bin_significand=floatbv_mult(
+    floatbv_typecast_exprt(bin_significand_int, round_to_zero_expr, float_type),
+    constant_float(1.192092896e-7, float_spec));
```

Comment:
Actually not all compilers support hexadecimal floating points, so I will have to revert to the non-hexadecimal representation.
For the table, a cleaner way to give it would be to recompute it each time, but since it involves relatively complicated operations and they are constants I'm not sure it's better.